pkg/client/unversioned/helper.go

```diff
@@ -258,16 +258,35 @@ func SetKubernetesDefaults(config *restclient.Config) error {
 	if config.APIPath == "" {
 		config.APIPath = legacyAPIPath
 	}
-	g, err := registered.Group(api.GroupName)
-	if err != nil {
-		return err
+	if config.GroupVersion == nil || config.GroupVersion.Group != api.GroupName {
+		g, err := registered.Group(api.GroupName)
+		if err != nil {
+			return err
+		}
+		copyGroupVersion := g.GroupVersion
+		config.GroupVersion = &copyGroupVersion
 	}
-	// TODO: Unconditionally set the config.Version, until we fix the config.
-	copyGroupVersion := g.GroupVersion
-	config.GroupVersion = &copyGroupVersion
 	if config.NegotiatedSerializer == nil {
 		config.NegotiatedSerializer = api.Codecs
 	}
-
 	return restclient.SetKubernetesDefaults(config)
 }
+
+func setGroupDefaults(groupName string, config *restclient.Config) error {
+	config.APIPath = defaultAPIPath
+	if config.UserAgent == "" {
+		config.UserAgent = restclient.DefaultKubernetesUserAgent()
+	}
+	if config.GroupVersion == nil || config.GroupVersion.Group != groupName {
```

Comment:
When we're creating a client [here](https://github.com/kubernetes/kubernetes/blob/ed3a29bd6aeb9874db4479e00fd2644176f01798/pkg/client/unversioned/helper.go#L49-L127) we are initiating all group clients at once, iow. batch, apps, rbac, etc. Currently `restclient.Config` has only one `GroupVersion` defined. This means that if I pass `batch/v2lpha1` I want only `batch` group to pick version `v2alpha1`. Others should use their defaults. Without this check I'll end up with a apps client trying to hit batch endpoints or others different combinations. As an example of such error see this listing, which was done without that check:

```
$ kubectl describe scheduledjob/pi
W0725 20:24:15.733731   19649 request.go:341] Field selector: batch/v2alpha1 - events - involvedObject.kind - ScheduledJob: need to check if this is versioned correctly.
W0725 20:24:15.733778   19649 request.go:341] Field selector: batch/v2alpha1 - events - involvedObject.name - pi: need to check if this is versioned correctly.
W0725 20:24:15.733783   19649 request.go:341] Field selector: batch/v2alpha1 - events - involvedObject.namespace - default: need to check if this is versioned correctly.
W0725 20:24:15.733790   19649 request.go:341] Field selector: batch/v2alpha1 - events - involvedObject.uid - aacdaafd-52a5-11e6-be6a-525400f25e34: need to check if this is versioned correctly.
Name:				pi
Namespace:			default
Schedule:			* * * * * ?
...
```